### PR TITLE
Optimize TensorList resizing.

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
This change optimizes the performance of `TensorList::Resize`
- simple inline functions are moved to the header
- `shared_ptr` in ShareData is now passed by value, allowing move semantics and reducing the number of atomic operations
- some code motion to improve inlining (e.g. wrapping frequent calls to DLL_PUBLIC functions into a trampoline function)

## Additional information:
Many of the changes were tuned experimentally. Don't hesitate to ask if you see something not obvious or outright weird.

### Affected modules and functionalities:
Buffer, Tensor, TensorList

### Tests:
No new functionality or functional changes - all existing tests apply
- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/data/buffer.h [hunk 1]
```diff
@@ -674,7 +674,10 @@ class DLL_PUBLIC Buffer {
   static double growth_factor_;
   static double shrink_threshold_;
 
-  static bool default_pinned();
+  static bool default_pinned() {
```
Comment (JanuszL) [inline dali/pipeline/data/buffer.h:677]: Can't you just move `RestrictPinnedMemUsage` definition here instead of adding `default_pinned` on top of it?

**Review by JanuszL — APPROVED** [review]

File: dali/pipeline/data/buffer.h (hunk 1 above)
Comment (mzient) [reply dali/pipeline/data/buffer.h:677]: I tried.... and it didn't help :upside_down_face:

File: dali/pipeline/data/tensor_list.cc [hunk 2]
```diff
@@ -744,30 +705,18 @@ void TensorList<Backend>::reserve(size_t bytes_per_sample, int batch_size) {
...
-  return state_.Get();
-}
-
-
 template <typename Backend>
 void TensorList<Backend>::recreate_views() {
   // precondition: type, shape are configured
   uint8_t *sample_ptr = static_cast<uint8_t *>(contiguous_buffer_.raw_mutable_data());
   int64_t num_samples = shape().num_samples();
+  auto &data_ptr = contiguous_buffer_.get_data_ptr();
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:713]: Hoisting this line was perhaps the biggest saving here.

File: dali/pipeline/data/tensor_list.cc [hunk 3]
```diff
@@ -744,30 +705,18 @@ void TensorList<Backend>::reserve(size_t bytes_per_sample, int batch_size) {
...
   for (int64_t i = 0; i < num_samples; i++) {
     // or any other way
     auto tensor_size = shape().tensor_size(i);
 
-    std::shared_ptr<void> sample_alias(contiguous_buffer_.get_data_ptr(), sample_ptr);
-    tensors_[i].ShareData(sample_alias, tensor_size * type_info().size(), is_pinned(), shape()[i],
+    tensors_[i].ShareData(std::shared_ptr<void>(data_ptr, sample_ptr),
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:718]: Having an intermediate variable and moving it was noticeably slower (but still noticeably faster than passing by const-ref and copying).

**Review by awolant — APPROVED** [review]

**Pull request merged**